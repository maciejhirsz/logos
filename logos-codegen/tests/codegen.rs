use std::{error::Error, io, path::PathBuf};

#[rstest::rstest]
#[case("simple")]
#[case("no_error_lut")]
pub fn test_codegen(#[case] fixture: &str) -> Result<(), Box<dyn Error>> {
    let mut fixture_dir = PathBuf::new();
    fixture_dir.push(env!("CARGO_MANIFEST_DIR"));
    fixture_dir.push("tests");
    fixture_dir.push("data");
    fixture_dir.push(fixture);

    let input = fixture_dir.join("input.rs");
    fixture_dir.push("output.rs");
    let output_file = fixture_dir;

    let input = std::fs::read_to_string(input)?;
    let output = std::fs::read_to_string(&output_file)?;

    let generated = logos_codegen::generate(input.parse()?);
    let generated = generated.to_string();

    if std::env::var("BLESS_CODEGEN").is_ok_and(|value| value == "1") {
        std::fs::write(&output_file, &generated)?;
        return Ok(());
    }

    assert_eq!(generated, output, "Codegen test failed: `{fixture}`, run tests again with env var `BLESS_CODEGEN=1` to bless these changes");

    Ok(())
}
