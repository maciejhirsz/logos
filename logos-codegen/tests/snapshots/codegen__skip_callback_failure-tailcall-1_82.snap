---
source: logos-codegen/tests/codegen.rs
expression: formatted
---
impl<'s> ::logos::Logos<'s> for Token {
    type Error = ();
    type Extras = ();
    type Source = str;
    fn lex(
        lex: &mut ::logos::Lexer<'s, Self>,
    ) -> std::option::Option<std::result::Result<Self, Self::Error>> {
        fn _logos_derive_compile_errors() {
            {
                compile_error!(
                    "Expected #[logos(skip(\"regex literal\"[, [callback = ] callback, priority = priority]))]"
                )
            }
            {
                compile_error!(
                    "Expected a named argument at this position\n\nhint: If you are trying to define a callback here use: callback = ..."
                )
            }
            { compile_error!("Expected: priority = <integer>") }
            { compile_error!("Expected an unsigned integer") }
            { compile_error!("Resetting previously set priority") }
            {
                compile_error!(
                    "Inline callbacks must use closure syntax with exactly one parameter"
                )
            }
            { compile_error!("Callback has been already set") }
            { compile_error!("Previous callback set here") }
            { compile_error!("Expected: callback = ...") }
            {
                compile_error!(
                    "Unknown nested attribute: unknown\n\nExpected: callback or priority"
                )
            }
            {
                compile_error!(
                    "Expected: #[logos(skip \"regex literal\")] or #[logos(skip(...))]"
                )
            }
            { compile_error!("expected literal") }
            {
                compile_error!(
                    "Expected #[logos(skip(\"regex literal\"[, [callback = ] callback, priority = priority]))]"
                )
            }
            { compile_error!("Expected a &str or &[u8] slice") }
            { compile_error!("Invalid nested attribute") }
            {
                compile_error!(
                    "The pattern #[regex(\"a\")] ::<skip> can match simultaneously with the following variants:\n  #[regex(\"a\")] ::<skip>\n  #[regex(\"a\")] ::A\n\n(all at the priority 2)"
                )
            }
            {
                compile_error!(
                    "The pattern #[regex(\"a\")] ::<skip> can match simultaneously with the following variants:\n  #[regex(\"a\")] ::<skip>\n  #[regex(\"a\")] ::A\n\n(all at the priority 2)"
                )
            }
            {
                compile_error!(
                    "The pattern #[regex(\"a\")] ::A can match simultaneously with the following variants:\n  #[regex(\"a\")] ::<skip>\n  #[regex(\"a\")] ::<skip>\n\n(all at the priority 2)"
                )
            }
        }
        unimplemented!()
    }
}
